packages/client/src/arpa_reporter/views/NewUpload.vue

```diff
@@ -1,111 +1,164 @@
-        ref="form"
-        @submit.prevent="onSubmit"
-      >
-        <div class="form-group">
-          <input
-            class="form-control-file"
-            type="file"
-            id="spreadsheet"
-            name="spreadsheet"
-            @change="changeFiles"
-            ref="files"
-          />
-        </div>
-        <div class="form-group">
-          <button
-            class="btn btn-primary"
-            type="submit"
-            :disabled="uploadDisabled"
-            @click.prevent="onSubmit"
-          >
-            {{ uploadButtonLabel }}
-          </button>
-          <a class="ml-3" href="#" @click="cancelUpload">Cancel</a>
-        </div>
-      </form>
-    </div>
+    <FormulateForm
+      name="new-upload"
+      :form-errors="formErrors"
+      v-model="values"
+      :schema="schema"
+      @submit="onSubmit"
+    >
+    <FormulateErrors
+      class="alert alert-danger"
+    />
+    </FormulateForm>
   </div>
 </template>
 
 <script>
+import Vue from 'vue';
+import VueFormulate from '@braid/vue-formulate';
 import { postForm } from '../store';
 
+Vue.use(VueFormulate, {
+  classes: {
+    prefix: 'formulate-',
+    outer: 'form-group',
+    input: (context, classes) => {
+      // set this to 'form-control' unless it's a file input
+      const inputClass = context.type === 'file' ? '' : 'form-control';
+      return [classes, inputClass].join(' ');
+    },
+    inputHasErrors: 'is-invalid',
+    help: 'form-text text-muted',
+    errors: 'list-unstyled text-danger',
+  },
+});
 export default {
-  name: 'NewUpload',
   data() {
     return {
-      error: null,
-      files: null,
-      uploading: false,
+      formErrors: [],
+      initialValues: () => ({
+        reportingPeriodId: this.$store.getters.viewPeriodID ?? '',
+        agencyId: this.$store.getters.user.agency.id,
+      }),
+
+      values: {},
+      schema: [
+        {
+          type: 'select',
+          name: 'reportingPeriodId',
+          label: 'Reporting Period',
+          placeholder: 'Select one',
+          options: this.reportingPeriodSelectItems(),
+          validation: 'required',
+        },
+        {
+          type: 'select',
+          name: 'agencyId',
+          label: 'Agency Code',
+          options: this.agencySelectItems(),
+          validation: 'required',
+        },
+        {
+          type: 'text',
+          label: 'Expenditure Code',
+          name: 'expenditure_code',
+          validation: 'required',
+        },
+        {
+          type: 'file',
+          label: 'Workbook File',
+          name: 'spreadsheet',
+          validation: 'required',
+          uploader: this.uploadFile,
+          // Don't upload the file until the form is submitted
+          uploadBehavior: 'delayed',
+        },
+        {
+          type: 'textarea',
+          label: 'Notes',
+          name: 'notes',
+        },
+        {
+          component: 'div',
+          class: 'row',
```

Comment:
I noticed there's a slightly wonky layout effect here.  I think the issue is that the `row` class isn't meant to be used without a corresponding `col`.

Probably the easiest fix is just to swap this for `d-flex`!
```suggestion
          class: 'd-flex',
```

Reply:
Ah, good catch!